### PR TITLE
add: topwallet and whitelist on celo

### PR DESCRIPTION
**Pull request by sirpy**
# Description
Support topwallet + whitelist on Celo

### CODE REVIEW

**Comment by johnsmith-gooddollar** [comment]
something wrong with the linter, need to re-check it carefully, build and test once i will finish review

**Review by johnsmith-gooddollar — APPROVED** [review]

**Comment by johnsmith-gooddollar** [comment]
@sirpy @L03TJ3   We have issues with initializing Celo wallet

1. there's no 'development-celo' network at `'@gooddollar/goodprotocol/releases/deployment.json'`. The same for QA, only prod section exists. ==> I've added CELO_FALLBACK_TO_PROD (default true) which enables using production values if nothing found on dev/qa env
2. there's no 'AdminWallet' address in 'production-celo' section. So celo wallet unable to initialize and app crashes (because address is undefined). ===> i've added CELO_ENABLED (default false) which disables celo from everywhere. So PR could be merged without affected dev/qa instances. But to support celo, we have to add admin wallet address and dev/qa network settings to goodprotocol

File: src/server/blockchain/Web3Wallet.js [hunk 1]
```diff
@@ -239,21 +209,20 @@ export class Wallet {
...
       }
-      // else log.warn('Failed adding admin wallet', { addr, balance, isAdminWallet, adminMinBalance })
 
       if (this.conf.env !== 'production') {
         const mainnetBalance = await this.mainnetWeb3.eth.getBalance(addr)
+
         if (parseFloat(web3Utils.fromWei(mainnetBalance, 'gwei')) > adminMinBalance * 100) {
           log.info(`admin wallet ${addr} mainnet balance ${mainnetBalance}`)
           this.mainnetAddresses.push(addr)
         }
```
Comment (sirpy) [inline src/server/blockchain/Web3Wallet.js:238]: this should go in the adminwallet

File: src/server/blockchain/CeloAdminWallet.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+const { celo, env, celoFallbackToProd } = conf
+let network = `${env}-celo`
+const defaultCeloGasPrice = (0.2 * 1e9).toFixed(0)
+
+// there's no "development/staging-celo" in deployment.json
+// added fallback to PROD settings, enables by the env flag
+if (!ContractsAddress.hasOwnProperty(network) && celoFallbackToProd) {
+  network = 'production-celo'
+}
+
```
Comment (sirpy) [inline src/server/blockchain/CeloAdminWallet.js:15]:
```suggestion
```

File: src/server/blockchain/CeloAdminWallet.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import ContractsAddress from '@gooddollar/goodprotocol/releases/deployment.json'
+
+import { Web3Wallet } from './Web3Wallet'
+import conf from '../server.config'
+
+const { celo, env, celoFallbackToProd } = conf
```
Comment (sirpy) [inline src/server/blockchain/CeloAdminWallet.js:6]:
```suggestion
const { celo, env } = conf
```

File: src/server/server.config.js [hunk 4]
```diff
@@ -590,6 +598,18 @@ const conf = convict({
     format: String,
     default: '',
     env: 'CF_WORKER_VERIFY_URL'
+  },
+  celoFallbackToProd: {
+    doc: 'Falls back to production-celo if no dev/qa settings in goodprotocol',
+    format: Boolean,
+    default: true,
+    env: 'CELO_FALLBACK_TO_PROD'
+  },
```
Comment (sirpy) [inline src/server/server.config.js:607]:
```suggestion
```

**Comment by sirpy** [comment]
@johnsmith-gooddollar install the beta goodprotocol package, fix tests

**Comment by johnsmith-gooddollar** [comment]
@sirpy i was checking latest beta (from master branch) of the protocol package - there was still 'production-celo' only network and no 'AdminWallet' address inside.

**Comment by johnsmith-gooddollar** [comment]
@sirpy Also, Lewis suggested me the address '0x3De7216149f12D8f51540D9A870149560FC11bfB'. I don't checked it too, will check today. If it works i will adjust env flags a bit:

CELO_CUSTOM_NETWORK=production_celo // could be set while no env settings in contracts, only prod
CELO_CUSTOM_ADMIN_ADDRESS=0x3De7216149f12D8f51540D9A870149560FC11bfB // could be set white no 'AdminWallet' address in celo contracts section

**Comment by sirpy** [comment]
> @sirpy Also, Lewis suggested me the address '0x3De7216149f12D8f51540D9A870149560FC11bfB'. I don't checked it too, will check today. If it works i will adjust env flags a bit:
> 
> CELO_CUSTOM_NETWORK=production_celo // could be set while no env settings in contracts, only prod CELO_CUSTOM_ADMIN_ADDRESS=0x3De7216149f12D8f51540D9A870149560FC11bfB // could be set white no 'AdminWallet' address in celo contracts section

no hacks please

**Comment by johnsmith-gooddollar** [comment]
@sirpy it won't work without them. 
the latest beta also won't fix the issue - i DMed in Slack

**Comment by johnsmith-gooddollar** [comment]
@sirpy thanks it works now
i removed all workarounds & pushed update

now will resolve conflicts then check tests

**Comment by johnsmith-gooddollar** [comment]
@L03TJ3 @sirpy 
Fixed everything expect blockchain tests failing because of 'AdminWallet contract low funds' / 'Returned values aren't valid, did it run Out of Gas?'
The same error happening on master branch and BC test also failing, so this looks not related to this PR changes

could you please check ? Thanks

**Review by johnsmith-gooddollar — APPROVED** [review]

**Comment by johnsmith-gooddollar** [comment]
@sirpy @L03TJ3 Also there's no test network for Celo. On AdminWallet we're using 'dapptest' / 'dapptest-mainnet', so i suppose the 'dattest-celo' should be added ? For now tests are using 'development' network (development-celo): https://github.com/GoodDollar/GoodServer/pull/385/files#diff-755919bd4c09bf52e7c7dcba502d0ecf2eca94857728bce80acd24906424aebaR5

File: .env.test [hunk 5]
```diff
@@ -3,7 +3,8 @@ PORT=3004
 NODE_ENV=test
 GUNDB_PASS=password
 GUNDB_SERVERMODE=true
-#MNEMONIC="myth like bonus scare over problem client lizard pioneer submit female collect"
+# NETWORK=fuse
+# MNEMONIC="season maximum wing sword benefit taxi tenant resist oyster bracket absurd firm"
```
Comment (johnsmith-gooddollar) [inline .env.test:7]:
```suggestion
# MNEMONIC="myth like bonus scare over problem client lizard pioneer submit female collect"
```

File: src/server/blockchain/MultiWallet.js [hunk 6]
```diff
@@ -83,7 +84,8 @@ class MultiWallet {
   }
 }
 
+//exclude celo wallet from tests
 export default new MultiWallet({
   122: AdminWallet, // "main" wallet goes first
-  42220: CeloAdminWallet
+  ...(conf.env === 'test' ? { 42220: CeloAdminWallet } : {})
```
Comment (johnsmith-gooddollar) [inline src/server/blockchain/MultiWallet.js:90]: @sirpy 
it doesn't helps as CeloAdminWallet have this.ready = this.init() at constructor
look over the commits history there was CELO_ENABLED flag i removed them
see how this was implemented. i made ready prop with getter and calling init() after first access instead from constructoer

Comment (johnsmith-gooddollar) [reply src/server/blockchain/MultiWallet.js]: @sirpy also you did mistake with condition - you're adding celo **if env is test**, should be vice versa
i've fixed: https://github.com/GoodDollar/GoodServer/pull/385/commits/cb931c5d200787c3f76b6acaa4e2e0303e8dd3b4

**Comment by johnsmith-gooddollar** [comment]
@sirpy could it be merged now ?

**Pull request merged**